src/audio/pipeline.c

```diff
@@ -177,8 +181,10 @@ static int pipeline_for_each_comp(struct comp_dev *current,
 
 		/* continue further */
 		if (ctx->comp_func) {
+			buffer->walking = true;
 			int err = ctx->comp_func(buffer_comp, buffer,
 						 ctx, dir);
+			buffer->walking = false;
```

Comment:
No following this.
1) All pipeline "walked" should be set to false prior to starting
2) Once started, we should only process paths where walked is false.
3) we should set path as walked = true after we have prcossed it.

Reply:
@lgirdwood actually, I was not quite willing to add this walking (or walked) flag, it introduces more traversing, let me summarize what we wanted to fix here:
![tplg](https://user-images.githubusercontent.com/10458135/83478704-196a4c80-a4c9-11ea-9995-104a642ed5d7.png)

The topology displayed above is quite typical for inter-connected pipelines, the BUF1.2 is quite special, it is used to connect the playback stream (pipe 1) and the capture stream (pipe 11), my PR here is aimed to add a negotiation step on this inter-connected buffer to make sure the sample format/rate is aligned on its source/sink components(muxdemux11.0 and smart_amp1.0).

With today's code the .params() is traversing like this:
a. if we are running .params() on the playback stream first, it will be propagated in this sequence:
PCM0P -- (BUF1.0) --> SMART_AMP1.0 -- (BUF1.1) --> SSP1 OUT DAI
The BUF1.2 is not traversed.
b. if we are running .params() on the capture stream first, it will be propagated in this sequence:
PCM0C -- (BUF11.0) --> MUXDEMUX11.0 -- (BUF11.1) --> SSP1 IN DAI
The BUF1.2 is not traversed neither.

What I am doing to fix the issue here is to add a special 'negotiate' step when the .params() is propagated to a component:
*If the component is connected to a inter-connected buffer, let's add the 'negotiate' step on this inter-connected buffer, before traversing forward with the existed sequence.*
How to judge if the buffer is inter-connected buffer? 
The buffer is an inter-connected buffer when it meet all items below:
1. its source and sink components are belonged to different pipelines.
2. the buffer can't be traversed in the .params() traversing of either its source and sink pipelines, as it is in the opposite side from its source/sink components traversing direction.
In our real design, if the item 2 above is met, the item 1 will be also met, so we only need to check the item 2 to decide if the buffer is an inter-connected buffer.

Reflected in the code, when traversing to a component in direction 'dir' (e.g. BUF1.0 --> SMART_AMP1.0), we will check if there is any other buffer (not the BUF1.0 which is the 'calling_buf' named here) in the opposite direction ('!dir') connected to this component, if yes (the BUF1.2), this is the inter-connected buffer, let's doing negotiate its params with the propagated one (and then return).

The check of 'other' buffer here can be done with simple "if (buffer != calling_buf)", but the suggestion to introduce a new 'walked' (or 'walking') flag will introduce a lot of work with maintaining to this flag if we want to design it like the DAPM path->walking flag, we are not clear about how to define this flag and what scenario it will be used, IMHO. The scenario I hit here is not for the full pipeline walking, but only for this multi-branch component specific only.

Reply:
So your conditional check is not actually checking for the true condition i.e. we want to check for a pipeline boundary buffer where one pipeline is active.
 
```c
/* is this buffer a boundary between two different pipelines ? */
if (buffer.pipeline != this.pipeline) {
    /* we can only negotiate params for this boundary buffer if one pipeline is active */
    /* otherwise we dont care as there is no cross pipeline data transfer */
    if (buffer.pipeline.state == ACTIVE) {
        /* negotiate params */
    }
}
```
This is self explanatory code that check for the real condition rather than a byproduct (which may not always evaluate as true in similar cases).

Reply:
> So your conditional check is not actually checking for the true condition i.e. we want to check for a pipeline boundary buffer where one pipeline is active.
> 
> ```c
> /* is this buffer a boundary between two different pipelines ? */
> if (buffer.pipeline != this.pipeline) {
>     /* we can only negotiate params for this boundary buffer if one pipeline is active */
>     /* otherwise we dont care as there is no cross pipeline data transfer */
>     if (buffer.pipeline.state == ACTIVE) {
>         /* negotiate params */
>     }
> }
> ```
> 
> This is self explanatory code that check for the real condition rather than a byproduct (which may not always evaluate as true in similar cases).

The boundary is only one (the first) of the 2 checks I mentioned above, for the example below:
```
host -> buf1.0->dmux->buf1.1->dai0
                  |-->buf2.0 ->dai1
```
though the buf2.0 is a boundary buffer, but we don't need a negotiate step for it as it has the same direction with the pipe 1 (so it doesn't meet the 2nd check).

Reply:
@keyonjie this patch is to check for walked buffers. It needs to have a per buffer flag to indicate teh walked state and this must be cleared prior to start of walk. There is no way around this.

Reply:
@keyonjie its only clear to me from this PR discussions that we
1) Search the current inactive pipline for active connected pipelines.
2) Assign params from connected active pipeline to our connection buffers.
3) ??
The implementation does not easily read as this flow.

Reply:
> @keyonjie its only clear to me from this PR discussions that we
> 
> 1. Search the current inactive pipline for active connected pipelines.
> 2. Assign params from connected active pipeline to our connection buffers.
> 3. ??
>    The implementation does not easily read as this flow.

Yes, it's not very easy to explain, especially in English. :)

Reply:
> @keyonjie this patch is to check for walked buffers. It needs to have a per buffer flag to indicate teh walked state and this must be cleared prior to start of walk. There is no way around this.

So I didn't want to use flag in my initial version, here with the latest version it actually be used as flag like "is_calling_buf", set it before walking further and clear it when return to the buffer, the negotiation will happen inside the ctx->comp_func(buffer_comp, buffer, ctx, dir) calling.

This is not perfect, I am not satisfy with it, but I don't want to introduce a DSPM path->walking liked flag, which could introduce much more complexity for us.

Reply:
I dont agree, adding use case specific changes is complexity. This should be generic like DAPM. The pipeline boundary negotiation should also be generic.